### PR TITLE
Better slice management for windows reads

### PR DESCRIPTION
**Pull request by jtattermusch**
Based on #17355.

Fixes https://github.com/grpc/grpc/issues/16464.

tcp_windows is much more basic implementation than tcp_posix (which supports resource quota, counters and a few other things missing in tcp_windows). 

The scenario from #16464:
1. there is a one-way stream of messages sending as fast as it can
2. for every read on windows, tcp_windows allocates a new slice (8KB in size) and but the actual size of the read is small, so there is a lot of allocated memory.
3. the slices only get disposed once the messages are processed by the C# layers.
4. there seems to be a large flow control window (because it's a high speed connection), so many messages will be in flight, each taking 8KB (regardless of its actual size).
5. In the example described in #16464 there is multiple channels, each one keeping a lot of slice around, to the memory usage grows like crazy and the thing eventually OOMs.

To fix, I basically used the same logic that is currently used by tcp_posix.cc -> a "leftover" buffer is kept between reads (buffer_trim is used to split the actual read data and the leftover), so all the slices allocated are fully utilized.
I've verified with an example that the memory consumption per channel is now much more reasonable (reduction from ~500MB to  ~35MB for ty example app).

Please refer  to tcp_posix when reviewing, as the logic used is almost identical
https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/tcp_posix.cc

### CODE REVIEW

**Comment by jtattermusch** [comment]
The fix is important enough to be backported to v1.17.x

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,020,410      Total (=)      2,020,410

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,174,076      Total (<)     11,174,080

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
I think it might be better to merge https://github.com/grpc/grpc/pull/17378 first and then upmerge to master, rather than merging in both branches? @muxi  @nicolasnoble

**Comment by muxi** [comment]
I thought upmerge is not performed until before next release, so merging both PR right now sounds better by helping reduce the conflicts when we upmerge later.

**Comment by jtattermusch** [comment]
@muxi I think that's wrong. The upmerge should be performed as soon as possible after stuff gets merged to a release branch (because if there's an important fix in the release branch, we can be 99.9% sure that the same fix will be needed in the master branch as well). I think upmerges should be frequent (which makes them simple) and should not be delayed unnecessarily.

I think merging the same PR twice is not a good idea in general (especially if the PR is non-trivial and has more than 1 commit), merging it just once to the release branch and then merging the branch with master makes it easier to track what fixes got merged where.

File: src/core/lib/iomgr/tcp_windows.cc [hunk 1]
```diff
@@ -112,7 +113,10 @@ typedef struct grpc_tcp {
 
   grpc_closure* read_cb;
   grpc_closure* write_cb;
-  grpc_slice read_slice;
+
+  /* garbage after the last read */
```
Comment (AspirinSJL) [inline src/core/lib/iomgr/tcp_windows.cc]: Not sure why it's called "garbage". But I'm okay with this since it's used elsewhere before this PR.

**Comment by jtattermusch** [comment]
The PR has already been merged into v1.17.x as https://github.com/grpc/grpc/pull/17378.
Because the review comments are only code-style changes, I'll just wait for the upmerge of #17378 to master and then fix the style in a followup PR on master.

**Comment by jtattermusch** [comment]
This was upmerged to master as part of #17423

**Pull request closed without merging**